middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml

```diff
@@ -836,78 +848,110 @@ let bigarray_set ~dbg ~unsafe kind layout b indexes value =
 
 (* Array accesses *)
 let array_access_validity_condition array array_kind index
-    ~(index_kind : L.array_index_kind) ~size_int =
+    ~(index_kind : L.array_index_kind) ~width_in_scalars ~size_int =
   [ multiple_word_array_access_validity_condition array ~size_int array_kind
-      index_kind ~width_in_scalars:1 ~index ]
+      index_kind ~width_in_scalars ~index ]
 
-let check_array_access ~dbg ~array array_kind ~index ~index_kind ~size_int
-    primitive : H.expr_primitive =
+let check_array_access ~dbg ~array (array_kind : P.Array_kind_for_length.t)
+    ~index ~index_kind ~size_int primitive : H.expr_primitive =
   checked_access ~primitive
     ~conditions:
       (array_access_validity_condition array array_kind index ~index_kind
-         ~size_int)
+         ~width_in_scalars:1 ~size_int)
     ~dbg
 
-let array_load_unsafe ~array ~index ~(mut : Lambda.mutable_flag)
-    (array_ref_kind : Array_ref_kind.t) ~current_region : H.expr_primitive =
+let array_load_unsafe ~array ~index ~(mut : Lambda.mutable_flag) array_kind
+    (array_ref_kind : Array_ref_kind.t) ~current_region : H.expr_primitive list
+    =
   let mut : Mutability.t =
     match mut with
     | Immutable | Immutable_unique -> Immutable
     | Mutable -> Mutable
   in
   match array_ref_kind with
-  | Immediates -> Binary (Array_load (Immediates, Scalar, mut), array, index)
-  | Values -> Binary (Array_load (Values, Scalar, mut), array, index)
+  | Immediates ->
+    [Binary (Array_load (array_kind, Immediates, mut), array, index)]
+  | Values -> [Binary (Array_load (array_kind, Values, mut), array, index)]
   | Naked_floats_to_be_boxed mode ->
```

Comment:
General comment: here and in other places, I'm wondering if we shouldn't split `Array_ref_kind.t` into a two-layer type, with `Naked_floats_to_be_boxed` in one case and something like `Normal_case` containing all the rest. I think this would also simplify the code of #3068 which already introduces the distinction, but keeps a full type and a no-floats-to-be-boxed case, instead of layering them.

Reply:
I've left a CR about this, will aim to do this once the remaining patches have been merged if it works out.